### PR TITLE
[Backport 2.x] Added validation for the new clusters field. (#633)

### PR DESCRIPTION
**Pull request by AWSHurneyt**
### Description
Manual cherry-pick of PR https://github.com/opensearch-project/common-utils/pull/633
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by engechas — DISMISSED** [review]

**Review by eirsep — APPROVED** [review]

File: src/main/kotlin/org/opensearch/commons/alerting/model/ClusterMetricsInput.kt [hunk 1]
```diff
@@ -31,35 +33,46 @@ data class ClusterMetricsInput(
...
+            if (clusters.isNotEmpty()) {
+                require(clusters.all { CLUSTER_NAME_REGEX.matches(it) }) {
+                    "Cluster names are not valid."
+                }
+            }
+
+            clusterMetricType = findApiType(constructedUri.path)
+            this.parseEmptyFields()
+        } catch (exception: Exception) {
+            throw CommonUtilsException.wrap(exception)
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/model/ClusterMetricsInput.kt:74]: plz add error log

Comment (eirsep) [reply src/main/kotlin/org/opensearch/commons/alerting/model/ClusterMetricsInput.kt:74]: why are we creating a common utils exception?
The plugin using common utils can handle the exception and wrap it as necessary?

File: src/main/kotlin/org/opensearch/commons/alerting/util/CommonUtilsException.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import org.opensearch.OpenSearchStatusException
+import org.opensearch.core.common.Strings
+import org.opensearch.core.rest.RestStatus
+import org.opensearch.index.IndexNotFoundException
+import org.opensearch.index.engine.VersionConflictEngineException
+import org.opensearch.indices.InvalidIndexNameException
+
+private val log = LogManager.getLogger(CommonUtilsException::class.java)
+
+class CommonUtilsException(message: String, val status: RestStatus, ex: Exception) : OpenSearchException(message, ex) {
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/util/CommonUtilsException.kt:20]: what benefit are we getting from removing the delegation of execption handling to the plugin? there is no business logic in common utils 
why should we have to throw a common utils exception?

File: src/main/kotlin/org/opensearch/commons/alerting/util/IndexUtils.kt [hunk 3]
```diff
@@ -11,6 +11,24 @@ import java.time.Instant
...
     companion object {
+        /**
+         * This regex asserts that the string:
+         *  The index does not start with an underscore _, hyphen -, or plus sign +
+         *  The index does not contain two consecutive periods (e.g., `..`)
+         *  The index does not contain any whitespace characters, commas, backslashes, forward slashes, asterisks,
+         *   question marks, double quotes, less than or greater than signs, pipes, colons, or periods.
+         *  The length of the index must be between 1 and 255 characters
+         */
+        val VALID_INDEX_NAME_REGEX = Regex("""^(?![_\-\+])(?!.*\.\.)[^\s,\\\/\*\?"<>|#:\.]{1,255}$""")
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/alerting/util/IndexUtils.kt:22]: why aren't we using opensearch core's index name validation?

File: src/main/kotlin/org/opensearch/commons/utils/ValidationHelpers.kt [hunk 4]
```diff
@@ -8,6 +8,24 @@ package org.opensearch.commons.utils
...
 import java.util.regex.Pattern
 
+/**
+ * This regex asserts that the string:
+ *  Starts with a lowercase letter, or digit
+ *  Contains a sequence of characters followed by an optional colon and another sequence of characters
+ *  The sequences of characters can include lowercase letters, uppercase letters, digits, underscores, or hyphens
+ *  The total length of the string can range from 1 to 255 characters
+ */
+val CLUSTER_NAME_REGEX = Regex("^(?=.{1,255}$)[a-z0-9]([a-zA-Z0-9_-]*:?[a-zA-Z0-9_-]*)$")
```
Comment (eirsep) [inline src/main/kotlin/org/opensearch/commons/utils/ValidationHelpers.kt:18]: why are we not re-using cross cluster validation from core?

File: src/main/kotlin/org/opensearch/commons/alerting/util/IndexUtils.kt (hunk 3 above)
Comment (AWSHurneyt) [reply src/main/kotlin/org/opensearch/commons/alerting/util/IndexUtils.kt:22]: This was just moved to common utils from alerting; it wasn't implemented as part of these changes. Core didn't have a regex for index names. This comment explains where the validation came from https://github.com/opensearch-project/alerting/pull/992/files#r1259175796

File: src/main/kotlin/org/opensearch/commons/alerting/util/CommonUtilsException.kt (hunk 2 above)
Comment (AWSHurneyt) [reply src/main/kotlin/org/opensearch/commons/alerting/util/CommonUtilsException.kt:20]: Removed `CommonUtilsException` in PR https://github.com/opensearch-project/common-utils/pull/639. Will cherry-pick that commit once that PR is merged.

File: src/main/kotlin/org/opensearch/commons/alerting/model/ClusterMetricsInput.kt (hunk 1 above)
Comment (AWSHurneyt) [reply src/main/kotlin/org/opensearch/commons/alerting/model/ClusterMetricsInput.kt:74]: Removed CommonUtilsException in PR https://github.com/opensearch-project/common-utils/pull/639. Will cherry-pick that commit once that PR is merged.

File: src/main/kotlin/org/opensearch/commons/utils/ValidationHelpers.kt (hunk 4 above)
Comment (AWSHurneyt) [reply src/main/kotlin/org/opensearch/commons/utils/ValidationHelpers.kt:18]: Discussed this in this comment
https://github.com/opensearch-project/common-utils/pull/633#discussion_r1563171254

**Review by jowg-amazon — APPROVED** [review]

**Review by eirsep — APPROVED** [review]

**Pull request merged**